omero/annotation_scripts/KeyVal_from_csv.py

```diff
@@ -0,0 +1,247 @@
+            img_name = row[0]
+            if( img_name not in dict_name_id ):
+                print("Can't find filename : {}".format(img_name) )
+            else:
+                img_ID = dict_name_id[img_name]         # look up the ID
+                img    = conn.getObject('Image',img_ID) # get the img
+
+                existing_kv = get_existing_map_annotations( img )
+                updated_kv  = copy.deepcopy(existing_kv)
+                print("Existing kv ")
+                for k,vset in existing_kv.items():
+                    print(type(vset),len(vset))
+                    for v in vset:
+                        print(k,v)
+
+                for i in range(1,len(row)):  # first entry is the filename
+                    key = header[i].strip()
+                    vals = row[i].strip().split(';')
+                    if( len(vals) > 0 ):
+                        for val in vals:
+                            if len(val)>0 : 
+                                if key not in updated_kv: updated_kv[key] = set()
+                                print("adding",key,val)
+                                updated_kv[key].add(val)
+
+                if( existing_kv != updated_kv ):
+                    nimg_updated = nimg_updated + 1
+                    print("The key-values pairs are different")
+                    remove_MapAnnotations( conn, 'Image', img.getId()  )
+                    map_ann = omero.gateway.MapAnnotationWrapper(conn)
+                    namespace = omero.constants.metadata.NSCLIENTMAPANNOTATION
+                    map_ann.setNs(namespace)
+                    # convert the ordered dict to a list of lists
+                    kv_list=[]
+                    for k,vset in updated_kv.items():
+                        for v in vset:
+                            kv_list.append( [k,v] )
+                    map_ann.setValue(kv_list)
+                    map_ann.save()
+                    img.linkAnnotation(map_ann)
+                else:
+                    print("No change change in kv's")
+
+    return "Added {} kv pairs to {}/{} files  ".format(len(header)-1,nimg_updated,len(dict_name_id))
+
+
+def run_script():
+
+    data_types = [rstring('Dataset')]
+    client = scripts.client(
+        'Add_Key_Val_from_csv',
+        """
+    This script processes a csv file, attached to a Dataset
+        """,
+        scripts.String(
+            "Data_Type", optional=False, grouping="1",
+            description="Choose source of images",
+            values=data_types, default="Dataset"),
+
+        scripts.List(
+            "IDs", optional=False, grouping="2",
+            description="Plate or Screen ID.").ofType(rlong(0)),
+
+        scripts.String(
+            "File_Annotation", grouping="3",
+            description="File ID containing metadata to populate."),
+
+        authors=["Christian Evenhuis"],
+        institutions=["MIF UTS"],
+        contact="christian.evenhuis@gmail.com"
```

Comment:
Update the authors and maybe institutions in all scripts to add yourself and OME team?
But certainly update the contact in all scripts, probably to `https://forum.image.sc/tag/omero`.  cc @joshmoore ?

Reply:
That tag is certainly fine thought we could also create an `omero-scripts` or similar if there's interest.